### ISSUE TITLE
Windows, SVG, possible font rendering deadlock under load

### ISSUE THREAD

**Issue by hallsbyra** [lovell/sharp#3535]
### Is this a possible bug in a feature of sharp, unrelated to installation?

- [x] Running `npm install sharp` completes without error.
- [x] Running `node -e "require('sharp')"` completes without error.

If you cannot confirm both of these, please open an [installation issue](https://github.com/lovell/sharp/issues/new?labels=installation&template=installation.md) instead.

### Are you using the latest version of sharp?

- [x] I am using the latest version of `sharp` as reported by `npm view sharp dist-tags.latest`.

If you cannot confirm this, please upgrade to the latest version and try again before opening an issue.

If you are using another package which depends on a version of `sharp` that is not the latest, please open an issue against that package instead.

### What is the output of running `npx envinfo --binaries --system --npmPackages=sharp --npmGlobalPackages=sharp`?

```
  System:
    OS: Windows 10 10.0.22621
    CPU: (24) x64 AMD Ryzen 9 3900X 12-Core Processor
    Memory: 31.84 GB / 63.93 GB
  Binaries:
    Node: 19.0.0 - C:\Program Files\nodejs\node.EXE
    npm: 8.19.2 - C:\Program Files\nodejs\npm.CMD
  npmPackages:
    sharp: ^0.31.3 => 0.31.3
```
### What are the steps to reproduce?

Sometimes our express server stops responding to requests. I traced it down to our usage of sharp. Sometimes, sharp returns a Promise that neither resolves nor rejects. I've isolated the problem into the code below that exhibits the bug. 
It appears to have something to do with SVG. I force the bug to happen quicker by running many operations in parallel.

### What is the expected behaviour?

The program should print `Generation 0..99` and then exit.

Instead, the program hangs, usually at the first or after just a few generations. The CPU usage of the process drops to 0.

### Please provide a minimal, standalone code sample, without other dependencies, that demonstrates this problem

```js
import sharp = require("sharp")

const svg = `<?xml version="1.0" standalone="no"?>
<svg xmlns="http://www.w3.org/2000/svg" xml:lang="en"
    height="846"
    width="540">
    <text
        text-anchor="middle"
        x="50%" y="209"
        font-size="28"
    >
    ${"ABCDEFG ".repeat(20)}
    ${"ABCDEFG ".repeat(20)}
    </text>
    <text
        text-anchor="middle"
        x="50%" y="237"
        font-size="28"
    >
    </text>
</svg>`


main()
    .then(() => console.log("OK"))
    .catch(err => console.log(`Error: ${err}`))

async function main() {
    for (let i = 0; i < 100; i++) {
        console.log(`Generation ${i}`)
        await Promise.all(Array(100).fill(1).map(_ => createImg()))
    }
}

async function createImg() {
    return await sharp({
        create: {
            width: 540,
            height: 846,
            background: { r: 255, g: 255, b: 255 },
            channels: 3,
        }
    })
    .composite([{ input: Buffer.from(svg) }])
    .toBuffer()
}
```

**Comment by lovell** [comment]
I cannot reproduce this locally. It sounds like a deadlock. Are you able to try older versions of Node.js? Does this only happen on Windows? What happens if the SVG does not contain text?

**Comment by hallsbyra** [comment]
> Are you able to try older versions of Node.js?

Yes. Problem persists in older versions of Node.js (see below).

> Does this only happen on Windows? 

Yes, it seems so (see below).

> What happens if the SVG does not contain text?

I think it must contain text. And it is sensitive to the exact text parameters. Changing the text elements makes it appear more/less often.

I've tested some different configurations:

```
Ubuntu/Node 19/Sharp 0.31.3 - Success
WSL/Alpine/Node 19/Sharp 0.31.3 - Success
WSL/Alpine/Node 18/Sharp 0.31.3 - Success
Windows/Node 19/Sharp 0.31.3 - Fail
Windows/Node 18/Sharp 0.31.3 - Fail
Windows/Node 16/Sharp 0.31.3 - Fail
Windows/Node 14/Sharp 0.31.3 - Fail
Windows/Node 19/Sharp 0.30.7 - Success*
Windows/Node 19/Sharp 0.31.0 - Fail
```
\* This is interesting! Suggests that the problem is introduced in 0.31.0

**Comment by lovell** [comment]
Thanks, this is useful extra information. This might be a fontconfig or pango problem. The `FC_DEBUG` environment variable can be used to make fontconfig provide debugging information at runtime, e.g. a value of 8191 will output *everything*.

https://www.freedesktop.org/software/fontconfig/fontconfig-user.html#DEBUG

Please can you try this and look at what is output just before the deadlock occurs. Is the output near/before the failure consistently the same on each run?

**Comment by hallsbyra** [comment]
Yes, that seems to be the case. Here are the last 10000 lines of two failing runs.

[last1.txt](https://github.com/lovell/sharp/files/10464995/last1.txt)
[last2.txt](https://github.com/lovell/sharp/files/10464996/last2.txt)

They are identical from line 7562 to the end.

**Comment by lovell** [comment]
Thank you, I don't see anything obviously wrong in these files (other than the use of UTF-16 :smirk: ).

If you haven't already done so, please can you get a process dump and inspect the call stack. You should be able to run `procdump` and `WinDbg` respectively to do this, there's some useful help at https://github.com/lowleveldesign/debug-recipes/blob/master/deadlocks/diagnosing-deadlocks.md

**Comment by hallsbyra** [comment]
Ok. Here's the output from `~*k`. WinDbg crashes when loading gsose.dll so I can't run the other commands. I'm not very used to WinDbg debugging.

[stacks.txt](https://github.com/lovell/sharp/files/10465758/stacks.txt)

**Comment by lovell** [comment]
Thanks, the `stacks.txt` file appears to be empty, please can you check.

**Comment by hallsbyra** [comment]
Sorry, fat fingers I guess. Try again!

**Comment by kleisauke** [comment]
I could not reproduce this hang, but it seems to be really slow on my Windows PC.

Creating a dummy font configuration like this:
`fonts.conf`:
```xml
<?xml version="1.0"?>
<!DOCTYPE fontconfig SYSTEM "urn:fontconfig:fonts.dtd">
<fontconfig>
  <!-- <dir>WINDOWSFONTDIR</dir> -->
  <!-- <dir>WINDOWSUSERFONTDIR</dir> -->
</fontconfig>
```

And running it with:
```cmd
$ set FONTCONFIG_PATH=%cd%
$ set FC_DEBUG=1024
$ node index.js
FC_DEBUG=1024
        Loading config file from C:\Users\kleisauke\Desktop\sharp-test\fonts.conf
        Loading config file from C:\Users\kleisauke\Desktop\sharp-test\fonts.conf done
Generation 0
Generation 1
<snip />
Generation 98
Generation 99
OK
```

Fixes this slowdown. Although, uncommenting those two `<dir>...</dir>` lines makes it slow again.

**Comment by lovell** [comment]
Great investigation skills, is is possible to narrow the problem down to either `WINDOWSFONTDIR` or `WINDOWSUSERFONTDIR`?

I wonder if fontconfig might be unable to cache fonts and would therefore scan for all fonts every time.

The sharp unit tests relating to font discovery on Windows occasionally timeout, and I suspect this might be related.

The fontconfig logic that converts `WINDOWSFONTDIR` and `WINDOWSUSERFONTDIR` into their real paths is in https://gitlab.freedesktop.org/fontconfig/fontconfig/-/blob/main/src/fcxml.c

**Comment by hallsbyra** [comment]
Leaving `WINDOWSFONTDIR` uncommented results in the bug. That's the only place I have fonts (`C:\Windows\Fonts`), so I suppose it means that it hangs when my fonts are loaded and works without any fonts loaded.

As a side note, when I run *without* any fonts (both lines commented away or removed), and with `FC_DEBUG=8191`, I get a crash very early with output ending abruptly like below. Don't know if this is a separate problem or related.

```
FcConfigSubstitute Pattern has 11 elts (size 16)
        family: "serif"(s)
        slant: 0(i)(s)
        weight: 80(f)(s)
        width: 100(i)(s)
        size: 12(f)(s)
        pixelsize: 12(f)(s)
        dpi: 72(f)(s)
        lang: "en"(s)
        prgname: "node.exe"(s)
        variable: DontCare(s)
        pangoversion: 15009(i)(s)

Best score 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
Sort Pattern has 25 elts (size 32)
        family: "Times New Roman"(s)
        familylang: "en"(s) "en-us"(w)
        stylelang: "en"(s) "en-us"(w)
        fullnamelang: "en"(s) "en-us"(w)
        slant: 0(i)(s)
        weight: 80(f)(s)
        width: 100(i)(s)
        size: 28(f)(s)
        pixelsize: 28(f)(s)
        hintstyle: 0(i)(s)
        hinting: False(s)
        verticallayout: False(s)
        autohint: False(s)
        globaladvance: True(s)
        hinting: False(s)
        verticallayout: False(s)
        autohint: False(s)
        globaladvance: True(s)
        dpi: 72(f)(s)
        lang: "en"(s)
        fontversion: 2147483647(i)(s)
        embeddedbitmap: True(s)
        decorative: False(s)
        namelang: "en"(s)
        prgname: "node.exe"(s)
        symbol: False(s)
        variable: DontCare(s)
        order: 0(i)(s)
        pangoversion: 15009(i)(s)

Null pattern
FcConfigS
```

**Comment by kleisauke** [comment]
I can confirm that this is indeed a deadlock/race condition issue. It is not caused by Pango (or Fontconfig), but Cairo seems to be the culprit now.

<details>
  <summary>Meme</summary>

![Cairo, Pango, FreeType](https://gitlab.gnome.org/GNOME/librsvg/uploads/d948dcd854d3f8e803b35cc76af39195/e36.jpg)
(image source: https://gitlab.gnome.org/GNOME/librsvg/-/merge_requests/288#note_708038)
</details>

Backporting commit https://gitlab.freedesktop.org/cairo/cairo/-/commit/76e0df566595a77e271de523b3d06f86f3e85813 and https://gitlab.freedesktop.org/cairo/cairo/-/commit/a8c1858cf2bb6efb35c0678d135fd522ece9e2a4 from the master branch of Cairo seems to fix this. I also backported commit https://gitlab.freedesktop.org/cairo/cairo/-/commit/531ec8d02db7d3482c3fb65d33dbb50c71b86999 and https://gitlab.freedesktop.org/cairo/cairo/-/commit/166d718099fc8898895d4ad2839d5fb9865a87a5 just to be sure.

@lovell Draft commit https://github.com/libvips/build-win64-mxe/commit/da49a60df2b4dc6b3be24a8f351df97623c43e3b incorporates this in the Windows builds. Let me know if I should make an 8.14.1-build2 release. :)

**Comment by lovell** [comment]
As if by magic there's a rather timely discussion about... wait for it... cairo v1.18.0, which will be the first stable release in almost 5 years. Perhaps doing nothing will be easier ;)

https://gitlab.freedesktop.org/cairo/cairo/-/issues/622#note_1732316

**Comment by lovell** [comment]
New cairo just dropped. https://github.com/lovell/sharp-libvips/commit/8a3f63edbd5d6a38167f18db42995acf2fc23415

**Comment by lovell** [comment]
v0.32.0 is now available with prebuilt binaries including the latest cairo (and pango). I'll close as these should fix the original problem, but please feel free to re-open with more details if the issue remains.

**Comment by hallsbyra** [comment]
I can confirm the problem is solved (at least for me).
